### PR TITLE
Port to TypeScript

### PR DESCRIPTION
**Pull request by chriseppstein**
Based on https://github.com/stefanpenner/template-ts

I've done this port on top of @SparshithNR's work to make use of the fs-merger. There's a followup to this PR in #191 that makes the dependency tracker work with the fs-merger. I've submitted these as separate PRs so we can keep the conversations distinct.

### CODE REVIEW

File: lib/strategies/persistent.ts [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+  /**
+   * By default initial dependencies are empty.
+   * @returns {Dependencies}
+   */
+  initialDependencies(srcDir, options) {
+    let result = this._syncCache!.get<string>('__dependencies__');
+    let dependencies;
+    if (result.isCached) {
+      /** @type {ReturnType<Dependencies['serialize']>} */
```
Comment (chriseppstein) [inline lib/strategies/persistent.ts:90]: I thought I removed all of these `@type` annotations. 🙃

File: tsconfig.json [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    "moduleResolution": "node",
+    "paths": {
+      "*": ["types/*"]
+    },
+    "sourceMap": true,
+    "strict": true,
+    "target": "es2018",
+    "types": ["node"],
+    "esModuleInterop": true,
+    "allowSyntheticDefaultImports": true
```
[the remark this answers is not in the document]
Comment (mike-north) [reply tsconfig.json:16]: If anything exposed through this library's public API only works b/c these options are enabled, all consumers of this library would have to set both of these to `true` or face compile errors in this lib's .d.ts files. 

I recommend leaving these as "false" in any project that exposes types to be consumed by others.

File: lib/strategies/persistent.ts (hunk 1 above)
Comment (stefanpenner) [reply lib/strategies/persistent.ts:90]: image

**Comment by chriseppstein** [comment]
@stefanpenner 

1. I pushed the change to convert `broccoli-node-api` to a dev dep.
2. The tests are not skipped anymore in #191. We shouldn't release this PR without also landing that one.
3. We also need @SparshithNR to review #191.

File: tsconfig.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (chriseppstein) [reply tsconfig.json]: I removed those options and changed the import statements to import/requires.

File: lib/strategies/persistent.ts (hunk 1 above)
Comment (chriseppstein) [reply lib/strategies/persistent.ts]: Ok. They're all removed now.

File: package.json [hunk 3]
```diff
@@ -34,7 +40,8 @@
   "dependencies": {
     "async-disk-cache": "^2.0.0",
     "async-promise-queue": "^1.0.3",
-    "broccoli-plugin": "^3.0.1",
+    "broccoli-node-api": "^1.7.0",
```
[the remark this answers is not in the document]
Comment (chriseppstein) [reply package.json]: This is a dev dep now.

**Comment by chriseppstein** [comment]
@stefanpenner I've addressed all your feedback.

File: index.ts [hunk 4]
```diff
@@ -158,35 +212,29 @@ module.exports = class Filter extends Plugin {
...
 
   async build() {
-    // @ts-ignore
     let srcDir = this.inputPaths[0];
-    // @ts-ignore
     let destDir = this.outputPath;
 
     if (this.dependencyInvalidation && !this.dependencies) {
-      this.dependencies = this.processor.initialDependencies(srcDir);
+      this.dependencies = this.processor.initialDependencies(srcDir, { fs: this.input });
```
Comment (SparshithNR) [inline index.ts:219]: We still have dependency over the `this.inputPaths`. We will not be able to remove this dependency, is it?

Comment (SparshithNR) [reply index.ts:219]: And dependency logic is more around considering `Filter` takes only one tree as input. #188  talks about an array of trees as an input. We probably have to revisit this part later to accommodate that change.

Comment (SparshithNR) [reply index.ts:219]: It looks like addressed in #191. Thanks

**Comment by chriseppstein** [comment]
This landed as part of #191

**Pull request closed without merging**